Update Identity Redis Samples (#30002)
Azure/azure-sdk-for-java: sdk/identity/azure-identity/src/samples/Azure-Cache-For-Redis/Jedis/Azure-AAD-Authentication-With-Jedis.md

### Table of contents

- [Prerequisites](#prerequisites)
- [Authenticate with Azure AD - Hello World](#authenticate-with-azure-ad-hello-world)
- [Authenticate with Azure AD - Handle Reauthentication](#authenticate-with-azure-ad-handle-reauthentication)
- [Authenticate with Azure AD - Using Token Cache](#authenticate-with-azure-ad-using-token-cache)
- [Authenticate with Azure AD - Azure Jedis Wrapper](#authenticate-with-azure-ad-azure-jedis-wrapper)
- [Troubleshooting](#troubleshooting)

#### Prerequisites

>  - Configuration of Role and Role Assignments is required before using the sample code in this document.

Author: The MS Doc is currently not available for this.
It will be up closer to public preview, so we can add the link then.

 - **Dependency Requirements:**
    ```xml
    <dependency>
        <groupId>com.azure</groupId>
        <artifactId>azure-identity</artifactId>
        <version>1.5.0</version>
    </dependency>
    
    <dependency>
        <groupId>redis.clients</groupId>
        <artifactId>jedis</artifactId>
        <version>4.0.1</version>
    </dependency>
    ```

#### Samples Guidance

> Familiarity with the [Jedis](https://www.javadoc.io/doc/redis.clients/jedis/latest/index.html) and [Azure Identity for Java](https://docs.microsoft.com/azure/developer/java/sdk/identity) client libraries is assumed.

Reviewer: We should also probably make this sentence a prereq instead of putting it here.

Integrate the logic in your application code to fetch an Azure AD access token via the Azure Identity library as shown below and replace it with the password configuring/retrieving logic in your application code.

```java
//Construct a Token Credential from Identity library, e.g. DefaultAzureCredential / ClientSecretCredential / Client CertificateCredential / ManagedIdentityCredential etc.
DefaultAzureCredential defaultAzureCredential = new DefaultAzureCredentialBuilder().build();

// Fetch an Azure AD token to be used for authentication. This token will be used as the password.
// Note: The Scopes parameter will change as the Azure AD Authentication support hits public preview and eventually GA's.
String token = defaultAzureCredential
    .getToken(new TokenRequestContext()
        .addScopes("https://*.cacheinfra.windows.net:10225/appid/.default")).block().getToken();

// SSL connection is required.
boolean useSsl = true;
// TODO: Replace Host Name with Azure Cache for Redis Host Name.
String cacheHostname = "<HOST_NAME>";

// Create Jedis client and connect to the Azure Cache for Redis over the TLS/SSL port using the access token as password.

> // Note, Redis Cache Host Name and Port are required below

Reviewer 2: If they're required why doesn't the API enforce that? If I was trying to use the API a comment like this would make me wonder what else I need to double check with documentation versus the API making me do the right thing.

Author: The requirement is for Azure AD scenarios, 
Since we're only building samples and not custom API in this case, we 're bounded by OSS library design.
So, as per discussion with Redis team, we decided to call out required parameters in comment as guidance for users.

Reviewer 2: Makes sense.

* [Client Certificate Credential](https://docs.microsoft.com/en-us/azure/developer/java/sdk/identity-service-principal-auth#client-certificate-credential)
* [Client Secret Credential](https://docs.microsoft.com/en-us/azure/developer/java/sdk/identity-service-principal-auth#client-secret-credential)
* [Managed Identity Credential](https://docs.microsoft.com/en-us/azure/developer/java/sdk/identity-azure-hosted-auth#managed-identity-credential)
* [Username Password Credential](https://docs.microsoft.com/en-us/azure/developer/java/sdk/identity-user-auth#username-password-credential)
* [Azure CLI Credential](https://docs.microsoft.com/en-us/azure/developer/java/sdk/identity-dev-env-auth#azure-cli-credential)
* [Interactive Browser Credential](https://docs.microsoft.com/en-us/azure/developer/java/sdk/identity-user-auth#interactive-browser-credential)
* [Device Code Credential](https://docs.microsoft.com/en-us/azure/developer/java/sdk/identity-user-auth#device-code-credential)

> 

Reviewer 2: same comment as Redis - use the type names here?

// Close the Jedis Client
jedis.close();

// Helper Code
private static Jedis createJedisClient(String cacheHostname, int port, String username, AccessToken accessToken, boolean useSsl) {
    return new Jedis(cacheHostname, port, DefaultJedisClientConfig.builder()
        .password(accessToken.getToken())
        .user(username)
        .ssl(useSsl)
        .build());
}

private static AccessToken getAccessToken(TokenCredential tokenCredential, TokenRequestContext trc) {
    return tokenCredential.getToken(trc).block();
}
```

#### Authenticate with Azure AD: Using Token Cache

> This sample is intended to assist in authenticating with Azure AD via Jedis client library. It focuses on displaying the logic required to fetch an Azure AD access token using a Token cache and to use it as password when setting up the Jedis instance. It also shows how to recreate and authenticate the Jedis instance using the cached access token when the client's connection is broken in error/exception scenarios. The Token Cache stores and proactively refreshes the Azure AD Access Token 2 minutes before expiry and ensures a non-expired token is available for use when the cache is accessed.

Reviewer: ```suggestion
This sample is intended to assist in authenticating with Azure AD via Jedis client library. It focuses on displaying the logic required to fetch an Azure AD access token using a token cache and to use it as password when setting up the Jedis instance. It also shows how to recreate and authenticate the Jedis instance using the cached access token when the client's connection is broken in error/exception scenarios. The token cache stores and proactively refreshes the Azure AD access token 2 minutes before expiry and ensures a non-expired token is available for use when the cache is accessed.
```

##### Migration Guidance
When migrating your existing your application code, you need to replace the password input with the Azure AD token.

> Integrate the logic in your application code to fetch an Azure AD access token via the Azure Identity library and store it in a Token Cache as shown below and replace it with the password configuring/retrieving logic in your application code.

Reviewer: ```suggestion
Integrate the logic in your application code to fetch an Azure AD access token via the Azure Identity library. Store the token in a token cache, as shown below. Replace the token with the password configuring/retrieving logic in your application code.
```

DefaultAzureCredential defaultAzureCredential = new DefaultAzureCredentialBuilder().build();

// Fetch an Azure AD token to be used for authentication. This token will be used as the password.
// Note: The Scopes parameter will change as the Azure AD Authentication support hits public preview and eventually GA's.
TokenRequestContext trc = new TokenRequestContext().addScopes("https://*.cacheinfra.windows.net:10225/appid/.default");
TokenRefreshCache tokenRefreshCache = new TokenRefreshCache(defaultAzureCredential, trc, Duration.ofMinutes(2));;
AccessToken accessToken = tokenRefreshCache.getAccessToken();

// SSL connection is required.
boolean useSsl = true;
// TODO: Replace <HOST_NAME> with Azure Cache for Redis Host name.
String cacheHostname = "<HOST_NAME>";

// Create Jedis client and connect to the Azure Cache for Redis over the TLS/SSL port using the access token as password.
// Note: Cache Host Name, Port, Username, Azure AD Access Token and ssl connections are required below.
Jedis jedis = createJedisClient(cacheHostname, 6380, "<USERNAME>", accessToken, useSsl);

int maxTries = 3;
int i = 0;

> while (i < maxTries) {

Reviewer 2: nit: can this be a for loop? Doing the increment manually just looks clunky.

    private volatile AccessToken accessToken;
    private final Duration refreshOffset;

    /**
     * Creates an instance of TokenRefreshCache
     * @param tokenCredential the token credential to be used for authentication.
     * @param tokenRequestContext the token request context to be used for authentication.
     * @param refreshOffset the refresh offset to use to proactively fetch a new access token before expiry time.
     */
    public TokenRefreshCache(TokenCredential tokenCredential, TokenRequestContext tokenRequestContext, Duration refreshOffset) {
        this.tokenCredential = tokenCredential;
        this.tokenRequestContext = tokenRequestContext;
        this.timer = new Timer();
        this.refreshOffset = refreshOffset;
    }

    /**
     * Gets the cached access token.
     * @return the AccessToken
     */
    public AccessToken getAccessToken() {
        if (accessToken != null) {
            return  accessToken;
        } else {
            TokenRefreshTask tokenRefreshTask = new TokenRefreshTask();
            accessToken = tokenCredential.getToken(tokenRequestContext).block();
            timer.schedule(tokenRefreshTask, getTokenRefreshDelay());

>             return accessToken;

Reviewer 2: Does something in here throw a reasonable exception or does this need to wrap and throw something meaningful?

Author: Timer, throws proper Exception types with cause as messages.

DefaultAzureCredential defaultAzureCredential = new DefaultAzureCredentialBuilder().build();

// Create Jedis Client using the builder as follows.
Jedis jedisClient = new AzureJedisClientBuilder()
    .cacheHostName("<HOST_NAME>") // TODO: Replace <HOST_NAME> with Azure Cache for Redis Host name.
    .port(6380) // Port is requried.
    .useSSL(true) // SSL Connection is required.
    .username("<USERNAME>") // Username is required.
    .credential(defaultAzureCredential) // A Token Credential is required to fetch Azure AD Access tokens.
    .build();

// Set a value against your key in the Redis cache.
jedisClient.set("Az:key", "sample");

// Close the Jedis Client
jedisClient.close();
```

#### Troubleshooting

##### Invalid Username Password Pair Error
In this error scenario, the username provided and the access token used as password are not compatible.
To mitigate this error, ensure that:
* On Portal, Under your `Redis Cache Resource` -> RBAC Rules, you've assigned the required role to your user/service principal identity.

> * On Portal, Under your `Redis Cache Resource` -> Advanced settings -> AAD Access Authorization box is checked/enabled, if not enable it and press the Save button.

Author: yes

##### Permissions not granted / NOPERM Error
In this error scenario, the authentication was successful, but your registered user/service principal is not granted the RBAC permission to perform the action.
To mitigate this error, ensure that:

> * On Portal, Under your `Redis Cache Resource` -> RBAC Rules, you've assigned the appropriate role (Owner, Contributor, Reader) to your user/service principal identity.

Author: Nope, under preview mode,  we have a seaparate tab called RBAC Rules.

    .port(6380) // Port is requried.
    .useSSL(true) // SSL Connection is required.
    .username("<USERNAME>") // Username is required.
    .credential(defaultAzureCredential) // A Token Credential is required to fetch Azure AD Access tokens.
    .build();

// Set a value against your key in the Redis cache.
jedisClient.set("Az:key", "sample");

// Close the Jedis Client
jedisClient.close();
```

#### Troubleshooting

##### Invalid Username Password Pair Error
In this error scenario, the username provided and the access token used as password are not compatible.
To mitigate this error, ensure that:
* On Portal, Under your `Redis Cache Resource` -> RBAC Rules, you've assigned the required role to your user/service principal identity.
* On Portal, Under your `Redis Cache Resource` -> Advanced settings -> AAD Access Authorization box is checked/enabled, if not enable it and press the Save button.

##### Permissions not granted / NOPERM Error
In this error scenario, the authentication was successful, but your registered user/service principal is not granted the RBAC permission to perform the action.

> To mitigate this error, ensure that:
> * On Portal, Under your `Redis Cache Resource` -> RBAC Rules, you've assigned the appropriate role (Owner, Contributor, Reader) to your user/service principal identity.
> * In the event you're using a custom role, then ensure the permissions granted under your custom role include the one required for your target action.

Reviewer: ```suggestion
To mitigate this error, navigate to your Azure Cache for Redis resource in the Azure portal. Confirm that:

* In **RBAC Rules**, you've assigned the appropriate role (Owner, Contributor, Reader) to your user/service principal identity.
* In the event you're using a custom role, ensure the permissions granted under your custom role include the one required for your target action.
```
